### PR TITLE
Changes

### PR DESCRIPTION
**Pull request by gundom0802**
Added in structure spell

### CODE REVIEW

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+		rotation = getConfigDataEnum("rotation", StructureRotation.class, StructureRotation.NONE);
+		mirror = getConfigDataEnum("mirror", Mirror.class, Mirror.NONE);
+		palette = getConfigDataInt("palette", 0);
+		integrity = getConfigDataFloat("integrity", 1);
+		includeEntities = getConfigDataBoolean("include-entities", false);
+
+		relativeOffset = getConfigDataVector("relative-offset", new Vector());
+
+		placeAtCaster = getConfigDataBoolean("place-at-caster", false);
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:46]: Name the option `point-blank` for consistency.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	public StructureSpell(MagicConfig config, String spellName) {
+		super(config, spellName);
+
+		structureKey = getConfigDataString("structure-key", "");
+
+		rotation = getConfigDataEnum("rotation", StructureRotation.class, StructureRotation.NONE);
+		mirror = getConfigDataEnum("mirror", Mirror.class, Mirror.NONE);
+		palette = getConfigDataInt("palette", 0);
+		integrity = getConfigDataFloat("integrity", 1);
+		includeEntities = getConfigDataBoolean("include-entities", false);
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:42]: This should probably default to true.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+			if (info.noTarget()) return noTarget(info);
+			data = info.spellData();
+		}
+
+		return castAtLocation(data);
+	}
+
+	@Override
+	public CastResult castAtLocation(SpellData data) {
+
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Remove unnecessary whitespace.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		NamespacedKey key = null;
+		try {
+			key = NamespacedKey.fromString(structureKey.get(data));
+		} catch (IllegalArgumentException ignore) {}
+		if (key == null) return new CastResult(PostCastAction.ALREADY_HANDLED, data);
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:70]: Remove the unnecessary null initial value, and return in the try-catch if there's an error instead of checking afterwards.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+		Structure structure = Bukkit.getStructureManager().loadStructure(key);
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:72]: Should this be `StructureManager#getStructure` instead? Not sure.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+		if (structure == null) {
+			return new CastResult(PostCastAction.ALREADY_HANDLED, data);
+		}
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Remove unnecessary whitespace and braces around the if.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+		Location target = data.location();
+		target.add(relativeOffset.get(data));
```
Comment (TonytheMacaroni) [inline core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:78]: That is not how relative offset is applied, use `Util#relativeOffset(Location, Vector)` (not the other one, as that's not actually a proper relative offset, just there for other spells that rely on broken behavior). Additionally, probably should add an absolute offset option as well.

**Review by TonytheMacaroni — CHANGES_REQUESTED** [review]
Couple things.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 4 above)
Comment (JasperLorelai) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:70]: The `NamespacedKey#fromString` throws `IllegalArgumentException` from some `Preconditions` but mainly returns `˙null` for invalid values.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 3 above)
Comment (JasperLorelai) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Fixed in the second commit.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 5 above)
Comment (JasperLorelai) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:72]: It calls that internally and if none is found it loads it in from the world (?) folder. I suggested loading, but I'm not too sure which would be better in this case either.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 4 above)
Comment (TonytheMacaroni) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:70]: Ah, I see. Disregard then.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 1 above)
Comment (gundom0802) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Changed in newest commit

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 2 above)
Comment (gundom0802) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Changed default value in newest commit

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 5 above)
Comment (gundom0802) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java:72]: I left this at loading. It seems like it has the same behavior but based on the documentation, loading seems like the better option here as it loads in from the folder itself.

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 6 above)
Comment (gundom0802) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Fixed in newest commit

File: core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java (hunk 7 above)
Comment (gundom0802) [reply core/src/main/java/com/nisovin/magicspells/spells/targeted/StructureSpell.java]: Redid the relative offset and added in absolute offset in newest commit. I mimic'd the OffsetLocation spell for the location code segment

File: core/src/main/java/com/nisovin/magicspells/castmodifiers/conditions/ClientNameCondition.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.entity.LivingEntity;
+import org.jetbrains.annotations.NotNull;
+
+import com.nisovin.magicspells.castmodifiers.Condition;
+
+public class ClientNameCondition extends Condition {
+
+	private String clientName;
```
Comment (Chronoken) [inline core/src/main/java/com/nisovin/magicspells/castmodifiers/conditions/ClientNameCondition.java:12]: Add a new line between the variable and the function.
